File: sorting-searching/sorting-algorithms/SIMD-Sorting/SIMDSorting.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class SIMDSorting {
    static final int BLOCK_SIZE = 4;

    static int[] mergeTwoSortedArrays(int[] left, int[] right) {
        int[] result = new int[left.length + right.length];

        int i = 0;
        int j = 0;
        int k = 0;

        while (i < left.length && j < right.length) {
            if (left[i] <= right[j]) {
                result[k++] = left[i++];
            } else {
                result[k++] = right[j++];
            }
        }

        while (i < left.length) {
            result[k++] = left[i++];
        }

        while (j < right.length) {
            result[k++] = right[j++];
        }

        return result;
    }

    static int[] simdStyleSort(int[] arr) {
        List<int[]> sortedBlocks = new ArrayList<>();

        for (int i = 0; i < arr.length; i += BLOCK_SIZE) {
            int end = Math.min(i + BLOCK_SIZE, arr.length);

            int[] block = Arrays.copyOfRange(arr, i, end);
            Arrays.sort(block);

            sortedBlocks.add(block);
        }

        while (sortedBlocks.size() > 1) {
            List<int[]> mergedBlocks = new ArrayList<>();

            for (int i = 0; i < sortedBlocks.size(); i += 2) {
                if (i + 1 < sortedBlocks.size()) {
                    mergedBlocks.add(
                        mergeTwoSortedArrays(sortedBlocks.get(i), sortedBlocks.get(i + 1))
                    );
                } else {
                    mergedBlocks.add(sortedBlocks.get(i));
                }
            }

            sortedBlocks = mergedBlocks;
        }

        return sortedBlocks.isEmpty() ? new int[]{} : sortedBlocks.get(0);
    }

    public static void main(String[] args) {
        int[] arr = {8, 3, 7, 4, 9, 2, 6, 1, 10, 5, 12, 11};

        System.out.println("Original array: " + Arrays.toString(arr));

        int[] sortedArray = simdStyleSort(arr);

        System.out.println("Sorted array: " + Arrays.toString(sortedArray));
    }
}

File: sorting-searching/sorting-algorithms/SIMD-Sorting/simd-sorting.cpp
```cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <immintrin.h>
using namespace std;

const int BLOCK_SIZE = 8;

void printArray(const vector<int>& arr) {
    for (int value : arr) {
        cout << value << " ";
    }
    cout << endl;
}

void sortBlockWithSIMDStyle(vector<int>& arr, int start, int end) {
    int length = end - start;

    if (length == BLOCK_SIZE) {
        __m256i values = _mm256_loadu_si256((__m256i*)&arr[start]);

        alignas(32) int temp[BLOCK_SIZE];
        _mm256_store_si256((__m256i*)temp, values);

        sort(temp, temp + BLOCK_SIZE);

        values = _mm256_load_si256((__m256i*)temp);
        _mm256_storeu_si256((__m256i*)&arr[start], values);
    } else {
        sort(arr.begin() + start, arr.begin() + end);
    }
}

vector<int> mergeTwoSortedArrays(const vector<int>& left, const vector<int>& right) {
    vector<int> result;

    int i = 0;
    int j = 0;

    while (i < left.size() && j < right.size()) {
        if (left[i] <= right[j]) {
            result.push_back(left[i]);
            i++;
        } else {
            result.push_back(right[j]);
            j++;
        }
    }

    while (i < left.size()) {
        result.push_back(left[i]);
        i++;
    }

    while (j < right.size()) {
        result.push_back(right[j]);
        j++;
    }

    return result;
}

void simdSorting(vector<int>& arr) {
    vector<vector<int>> sortedBlocks;

    for (int i = 0; i < arr.size(); i += BLOCK_SIZE) {
        int end = min(i + BLOCK_SIZE, (int)arr.size());

        sortBlockWithSIMDStyle(arr, i, end);

        vector<int> block(arr.begin() + i, arr.begin() + end);
        sortedBlocks.push_back(block);
    }

    while (sortedBlocks.size() > 1) {
        vector<vector<int>> mergedBlocks;

        for (int i = 0; i < sortedBlocks.size(); i += 2) {
            if (i + 1 < sortedBlocks.size()) {
                mergedBlocks.push_back(
                    mergeTwoSortedArrays(sortedBlocks[i], sortedBlocks[i + 1])
                );
            } else {
                mergedBlocks.push_back(sortedBlocks[i]);
            }
        }

        sortedBlocks = mergedBlocks;
    }

    if (!sortedBlocks.empty()) {
        arr = sortedBlocks[0];
    }
}

int main() {
    vector<int> arr = {8, 3, 7, 4, 9, 2, 6, 1, 10, 5, 12, 11};

    cout << "Original array: ";
    printArray(arr);

    simdSorting(arr);

    cout << "Sorted array: ";
    printArray(arr);

    return 0;
}
